sdk/core/allocator/main.cc

```diff
@@ -250,14 +250,9 @@ namespace
 				           "allocation, sleeping",
 				           bytes);
 				// Drop the lock while yielding
+				auto expected = ++freeFutex;
```

Comment:
Hmm, good point.  I made this change because I suspected that the old version was the cause of a test failure, but that turned out to be a scheduler bug, but the newer version is just differently wrong.

I am not sure if there is a way of avoiding that other than to build a proper condition variable that's protected with a lock.  Given that we expect to have fewer than 32 threads in the common case, we possibly could do `freeFutex |= (1 << thread_id_get_fast())`, but that won't work with more than 32 threads (I suppose we could have multiple futex words for that case).

Any suggestions?

Reply:
This is now fixed as per our side discussion.  The value in the wait futex is not the number of waiters, it is the amount of free memory.  A value of -1 indicates no waiters.  If we free things, we wake waiters if there are any.  If we sleep, there are two possible races:

 - We set an expected amount of memory and more memory is consumed.  We will spuriously wake and retry here.  That's mildly annoying for performance (though in a fairly niche case) but not a correctness bug.
 - We set an expected amount of memory and a sequence of allocations and frees results in the same amount of memory. We now sleep.  This is a small correctness bug because it's possible that the new shape of the heap would allow the allocator to make progress for this allocation, even though the amount of memory is the same, but it's a bit unlikely.

Reply:
A thought after our discussion: if we're sleeping waiting for a deallocation, then the size we're requesting must be larger than the largest free block at the time.  So let's sleep on that number rather than the number of free bytes.  That reduces the first case to someone allocating the largest block, not merely some block, and closes the second case, yes?

Reply:
Do we have a cheap way of finding that?  The amount of free space is always tracked.  I think blocks are sorted by size, so maybe we can find the largest size easily?

Reply:
The tries are not exactly sorted (see the big block comment at https://github.com/microsoft/cheriot-rtos/blob/main/sdk/core/allocator/alloc.h#L496-L548 that really belongs somewhere else) so it's O(log n) to find the maximum size in a tree bin (it's somewhere on the path to the rightmost *leaf* node), but maybe that's OK.  It's O(1) to increase the maximum when coalescing, and we only need to do the O(log n) when the largest chunk gets allocated or split up.  Needs more thought; let's leave it as an issue for a later PR.